Fix recursion issue with nested instances and unions

Peilonrayz opened the pull request:
# Description

Fixes #9657

## Test Plan

I have added three new tests to `test-data/unit/check-unions.test`

-   `testNestedProtocolUnions`
    This is an exact copy of the MVCE I included in the issue.

-   `testNestedProtocolGenericUnions`
    At first I 'fixed' the issue by not running the `for` if `left` is a protocol in `mypy.subtypes:is_protocol_implementation`. I however noticed that this would still cause the same issue if we changed to `Iter = Union[Iterator[T], List[T]]`. And all I'd have done is add a subtle bug. This is just the same as `testNestedProtocolUnions` but to account for non-protocols.

-   `testNestedProtocolGenericUnionsDeep`
    This tests that we can go 5 `Iter`s deep. This is important as we can fix the bug at 3 levels by adding the following to `mypy.sametypes:is_same_type`.

    ```python
    if isinstance(left, UnionType) is not isinstance(right, UnionType):
        return False
    ```

My changes pass all the tests ran via `./runtests.py`.

## Origin of the Bug

The issue comes from [assuming in `mypy.subtypes:is_protocol_implementation`](https://github.com/python/mypy/blob/985a20d87eb3a516ff4457041a77026b4c6bd784/mypy/subtypes.py#L525).

1.  The assumed type is an instance that has a union as its argument.
2.  `mypy.sametypes:is_same_type` works fine for the `Instance`.
3.  We check the `Union`s are the same type, again with `mypy.sametypes:is_same_type`.
4.  This falls apart as we then call `simplify_union` which goes on to call `mypy.subtypes:is_protocol_implementation`.

    In this second call to `is_protocol_implementation` what we assume the type to be stays the same. This causes problems because we're constantly simplifying what we assume the type to be against itself.

To circumvent this semi-infinite loop I've forced the left value to always be assumed to be a simplified union. This is risky as this is a new assumption. It seems to work, but I'm unsure if this is because of a lack of tests that'd put a non-simplified union on the left side.

## Alternates

I can think of two different solutions that don't add an assumption to `mypy.sametypes:is_same_type`.

1.  Adjusting the assumed types in `mypy.subtypes:is_protocol_implementation` to correctly handle the recursion. I don't know how to implement this.
2.  Changing UnionType to have a 'simplified' attribute. I could implement this. But it'd be better wherever the assumptions are being made, and I do not know where that is.

JukkaL commented on mypy/sametypes.py:
```diff
@@ -25,7 +25,6 @@ def is_same_type(left: Type, right: Type) -> bool:
         # types can be simplified to non-union types such as Union[int, bool]
         # -> int. It would be nice if we always had simplified union types but
         # this is currently not the case, though it often is.
-        left = simplify_union(left)
```
Unfortunately this doesn't work for cases like checking whether `Union[A, B]` is the same type as `Union[B, A]`, where `B` is a subclass of `A`.

One possible way to fix this would be to use a simpler replacement for `is_same_type` in `is_protocol_implementation`. Maybe we can just use type equality (==)?

Peilonrayz replied:
Thank you for the feedback!

Ah, my assumption is indeed wrong. Unfortunately I don't know how I could integrate that as a test for mypy. Could you help me get one, so I can be sure any other changes I make don't regress this?

Oh, I previously tried something similar - copying `is_same_type` to `subtypes` with my change but that didn't work. However just using `==` seems to work fine. I'll commit this soon.

JukkaL replied:
You can write unit tests for `is_same_type` in `mypy/test/testtypes.py`. `TypeFixture` contains types that you can use as union type items.

Peilonrayz replied:
Thank you. I should be able to get a new unit test committed tomorrow.

JukkaL reviewed (changes requested):
Thanks for the PR and the detailed explanation! 

This seems to regress some supported use cases. I left an alternative idea below.

Peilonrayz commented:
As suggested by JukkaL, I have changed from `mypy.sametypes.is_same_type(l, left)` to `l == left`. And it seems to work fine.

I have added three additional tests to check if `get_proper_type` or `mypy.sametypes.simplify_union` is needed. These tests pass so I can't see any issues here.

JukkaL reviewed (approved):
Thanks for the updates! Looks good now. It would be good to if you can add a unit test for `is_same_type` (but not necessary), as explained in my previous comment.

Peilonrayz commented:
Sorry for the delay. I have now added the test in what I think is the best place in the file. I ensured that the test fails on the older solution locally. I hope this is the correct place to add the test, and thank you for helping me find the file in the first place.

hauntsaninja commented:
Thanks for a well-explained issue report and subsequent fix to a tricky bug!
